### PR TITLE
Adding flatpak_remote module

### PR DESCRIPTION
**Pull request by JayKayy**
##### SUMMARY

This pull request adds the `flatpak_mote` module which is used to manage flatpak remotes or repos on a given host. This will allow users to configure their flatpak installs and upstream remotes for installing flatpaks.

##### ISSUE TYPE
 - New Module Pull Request

##### ANSIBLE VERSION
```
ansible 2.4.3.0
  config file = /home/jkwiatko/.ansible.cfg
  configured module search path = [u'/home/jkwiatko/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Mar 14 2018, 13:36:31) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]


```

##### ADDITIONAL INFORMATION

```
[jkwiatko ansible]$ sudo flatpak remote-list
Name                            Options            
flathub                         system             
flathub-1                       system             
gnome-nightly                   system             
kderuntime                      system             
com.spotify.Client-origin       system,no-enumerate
com.vinszent.GnomeTwitch-origin system,no-enumerate
org.gnome.gedit-origin          system,no-enumerate
[jkwiatko ansible]$ ./hacking/test-module -m ./lib/ansible/modules/packaging/os/flatpak_repo.py -a 'name=https://sdk.gnome.org/gnome.flatpakrepo state=present'
* including generated source, if any, saving to: /home/jkwiatko/.ansible_module_generated
* ansiballz module detected; extracted module source to: /home/jkwiatko/debug_dir
***********************************
RAW OUTPUT

{"invocation": {"module_args": {"state": "present", "name": "https://sdk.gnome.org/gnome.flatpakrepo"}}, "changed": false}


***********************************
PARSED OUTPUT
{
    "changed": false, 
    "invocation": {
        "module_args": {
            "name": "https://sdk.gnome.org/gnome.flatpakrepo", 
            "state": "present"
        }
    }
}
[jkwiatko ansible]$ sudo flatpak remote-list
Name                            Options            
flathub                         system             
flathub-1                       system             
gnome                           system             
gnome-nightly                   system             
kderuntime                      system             
com.spotify.Client-origin       system,no-enumerate
com.vinszent.GnomeTwitch-origin system,no-enumerate
org.gnome.gedit-origin          system,no-enumerate
[jkwiatko ansible]$ ./hacking/test-module -m ./lib/ansible/modules/packaging/os/flatpak_repo.py -a 'name=https://sdk.gnome.org/gnome.flatpakrepo state=absent'
* including generated source, if any, saving to: /home/jkwiatko/.ansible_module_generated
* ansiballz module detected; extracted module source to: /home/jkwiatko/debug_dir
***********************************
RAW OUTPUT

{"invocation": {"module_args": {"state": "absent", "name": "https://sdk.gnome.org/gnome.flatpakrepo"}}, "changed": true}


***********************************
PARSED OUTPUT
{
    "changed": true, 
    "invocation": {
        "module_args": {
            "name": "https://sdk.gnome.org/gnome.flatpakrepo", 
            "state": "absent"
        }
    }
}

```

### CODE REVIEW

**Comment by JayKayy** [comment]
@dagwieers

**Comment by oolongbrothers** [comment]
Hello again!

After looking at the `flatpak` module, I got interested in this companion `flatpak_remote` module and tested around with it.

To start with, this was in a bit of a messy state so I cleaned it up a bit and found it working fine after that. Great effort!

Wasn't sure what would be the best way to send you my changes. What I did is that I created a branch forked from ansible `devel` and then merging in your branch (sorry, couldn't get rebase to behave there): https://github.com/oolongbrothers/ansible/tree/testing_PR28808_clean_branch

Would you be so kind and have a look at my commits? I tried to explain everything in the commit messages.

If you like some of the stuff, I guess it would be cleanest if you:

* First pulled my branch (`git pull git@github.com:oolongbrothers/ansible.git testing_PR28808_clean_branch`) into a new local branch
*  Then cherry-picked the ones you want into your original branch

Regards,
//oolongbrothers

**Comment by JayKayy** [comment]
@oolongbrothers 
I have gotten the base functionality down. Thanks for your feedback. However, I am seeing weird behavior as the `flatpak remote-list --user -d` will only display an arbitrary amount of characters of the remote's url. so I still need to find out how to use the cli to get the full url without the columns truncating the end of it. I will fix the auto test verification shortly just wanted to get the functionality back in.

**Comment by oolongbrothers** [comment]
Wow, thanks! Great to see that you pushed the module even a bit further along. I like many of the improvements.

I tested with your latest version. I found some smaller bugs that I have fixes for, more on that later.

I had a look at what you meant with the URL being truncated. I think I have en explanation. It's not actually truncating, it's more of a misconception on our side. We currently assume that the URL for the remote (as listed in remote-list) is the same as the `.flatpakrepo` name.
In reality, these URLs do not have to have anything to do with each other. When installing a remote from a `.flatpakrepo` file, the URL for the remote is taken from the `Url` property in the file and can point to some other place. For an example, look at this `.flatpakrepo` file: https://distribute.kde.org/kdeapps.flatpakrepo, it points to http://distribute.kde.org/flatpak-apps-testing/ as the remote URL.

Ultimately, if we would want to properly support updating the URL of existing remotes, we would need to actually download the provided `.flatpakrepo` file and check the `Url` property, just to verify the existing remote has the correct URL. Also, there are other properties in the file, like `GPGKey` that we then also might want to support. When we do all that, we would pretty much have support for adding flatpak-remotes manually as well (i.e. not from a `.flatpakrepo` file). I am not sure if this is actually worth it right now. Maybe we would want to add such complex features later-on and concentrate on just getting the module ready for inclusion. My proposal is for now to simplify the matching logic to only check for remote name.

I can push some commits to a new branch on my fork (this time without merge) to illustrate what I mean. I will also add the bugfixes there. More explanations in the commit messages, as before.

By the way, I have a test playbook for verifying this proposed module's behaviour, in case you are interested: https://gist.github.com/oolongbrothers/41842e2cfc8826ae7bf8b1b6168ad43e

**Comment by JayKayy** [comment]
@oolongbrothers Can you rebase your branch with ansible devel? When i go to compare our branches changes it says that theres 5,000+ changed files. Also I could just be using github correctly. Its difficult for me to see the specific changes in place for the flatpak_remote.py file.

**Comment by oolongbrothers** [comment]
Oops, sorry! My bad, I had not properly synced from upstream.
However, I just rebased from upstream devel and force-pushed https://github.com/oolongbrothers/ansible/tree/testing_PR28808. Does it look better now for you too?

Sorry that it took so long, I had a lot going on lately. But I should be rather attentive over the coming Easter weekend in case anything else comes up.

**Comment by oolongbrothers** [comment]
I tested the module again and had another look at the code and I am pretty happy with it now.

I did just push some small changes to my branch: Some wording improvement for the docs and adding  my name to the copyright notice. Just gold-plating now, really.

I think it's about time that some maintainer would have a look at it and give us some feedback :)

**Comment by JayKayy** [comment]
:) thanks @oolongbrothers ill pull that in this evening.

**Comment by JayKayy** [comment]
@oolongbrothers Are your changes in a particular branch? when i click on your commit it sends me to an odd tree or commit.

**Comment by oolongbrothers** [comment]
Oh no, did I do something wrong again this time? I'm sorry. I tried to do it exactly like last time and like the Ansible documentation says. But it doesn't say anything about this triangle situation we have, so my workflow seems not to be perfect.

Just now I rebased on upstream devel again and pushed to a new branch `testing_PR28808_new_rebase`. It can be found here: https://github.com/oolongbrothers/ansible/tree/testing_PR28808_new_rebase

The two commits I pushed are those:
https://github.com/oolongbrothers/ansible/commit/ff67d0a3566a491db362a3c7ca1f6ffb8630bf94
https://github.com/oolongbrothers/ansible/commit/5d5ee0c71162626df63580fe4a7f21f83798aa34

Do these look O.K. for you now, @JayKayy ?

**Comment by JayKayy** [comment]
I dont know how many times i need to rebase this branch. I cant get that merge commit out of the history? @oolongbrothers any thoughts?

**Comment by SISheogorath** [comment]
i would actually do the following: 

```sh
# copy your changes
cp  lib/ansible/modules/packaging/os/flatpak_remote.py  lib/ansible/modules/packaging/os/flatpak_remote.py.new
# switch to master
git checkout master
# given the case you followed the github guide and added an upstream branch for ansible/ansible
git pull upstream master
# Delete the branch
git branch -D flatpak-repo
# recreate it based on master
git checkout -b flatpak-repo
# move your changes back in place
mv lib/ansible/modules/packaging/os/flatpak_remote.py.new lib/ansible/modules/packaging/os/flatpak_remote.py
# add them
git add lib/ansible/modules/packaging/os/flatpak_remote.py
# commit them
git commit
# overwrite your branch on github
git push -f --set-upstream origin flatpak-repo
```

As you have to redo your commit, check that you contain all your *useful* commit messages.

You can also try to do an interactive rebasing which allows you to omit commits, mark others as fixup to merge them into the last one before, … but for this long history I wouldn't recommend that.

**Comment by oolongbrothers** [comment]
Hm, a bit sad that one can't do a git annotate any more to see what line came from what commit and whom... But eh, go for it!
It's more important that the branch is in shape for a merge to devel!

**Comment by SISheogorath** [comment]
Oh ehm, right, when I gave the advice I didn't took that into account :x Sorry for that.

I just search for such a module and saw there is help needed so I went for some. Look like it wasn't completely perfect but when all contributors are fine with it, I really hope it's not a problem and flatpak makes it into the devel and soon stable line :)

**Comment by oolongbrothers** [comment]
@SISheogorath No worries! I don't either know a great way to work in a triangle of a project and two forks towards one pull request. A clean way might be to cherry-pick every commit. But that is quite cumbersome as well.
Great to hear you are interested in this module as well.

No need to redo anything, @JayKayy .

The branch seems to be in good shape for a merge to devel now, let's hope that it's not far off now any more :smiley:

**Comment by JayKayy** [comment]
@oolongbrothers thank you for all the help and contributions. I'm looking forward to it going some traction!

File: lib/ansible/modules/packaging/os/flatpak_remote.py
Comment (dagwieers) [inline lib/ansible/modules/packaging/os/flatpak_remote.py:39]: If a parameter is not required, you don't need to add this.
It helps readability if only required parameters have the required option.

File: lib/ansible/modules/packaging/os/flatpak_remote.py
Comment (dagwieers) [inline lib/ansible/modules/packaging/os/flatpak_remote.py:117]: IMO it is better to move this test further up in **_flatpak_command()**.
And leave the module through the final **exit_json()** just as if you would if you're not in check_mode.

File: lib/ansible/modules/packaging/os/flatpak_remote.py
Comment (dagwieers) [inline lib/ansible/modules/packaging/os/flatpak_remote.py:153]: Why don't we return the return_code back to the user ?
I think we should also return the command output back to the user, especially if the command failed.

File: lib/ansible/modules/packaging/os/flatpak_remote.py
Comment (dagwieers) [inline lib/ansible/modules/packaging/os/flatpak_remote.py:238]: I would rewrite this as :
```python
    if result != 0:
        module.fail_json(msg=output, changed=changed)

    module.exit_json(changed=changed)
```

**Review by dagwieers — CHANGES_REQUESTED** [review]
So I would do this slightly different and more in line with what other modules do that run command line tools.

- Create a global result dictionary with changed=False
- If the command fails, add result['rc'] and result['stdout'] and/or result['stderr']
- Return the complete commandline in result['command']
- Call fail_json and exit_json with **result added

The module will be more robust, easier to read and more concise.
And if it fails, the user will have more means to understand why.

**Comment by oolongbrothers** [comment]
Thank you very much for you feedback, @dagwieers !

I'm very much on board with improving user feedback and consistency between modules.

I did an attempt at incorporating your feedback with this commit to my own form: https://github.com/oolongbrothers/ansible/commit/079cd5945cb1d9094ff0d96eee31d63bcb3084ff

Is this what you meant to convey? I was looking for a good example module for inspiration, but I found it hard to find one fitting your description. So I did what I felt was right based on your comments.

Let us know what you think!

**Comment by dagwieers** [comment]
@oolongbrothers A lot better IMO. Code is more concise. I have some more remarks added to your commit to simplify the code even more.

**Comment by dagwieers** [comment]
Beware that the deadline for getting new features/modules accepted in [Ansible v2.6](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_6.rst#release-schedule) is nearing, it is set to either 2018-05-17 or 2018-05-25. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-devel or add a comment to [the Public Ansible Project meeting agenda](https://github.com/ansible/community/issues/308) to get it resolved.

**Comment by oolongbrothers** [comment]
Thank you once more for your feedback, @dagwieers. That's very valuable.
I just pushed a new commit addressing your latest points. https://github.com/oolongbrothers/ansible/commit/9b4f4fd6b4b81a75b23603e42dd2fe711b1726ac

**Comment by oolongbrothers** [comment]
Also, thank you for the heads-up regarding the 2.6 code freeze, @dagwieers . I will do my best to do anything I can to get this in (and maybe the `flatpak` module). Are you up for it, @JayKayy ? :smiley:

**Comment by oolongbrothers** [comment]
I just did some improvements to the module here: https://github.com/oolongbrothers/ansible/commit/97ac821b77510b9857d3ccdbe3f1443b0d5c35f5
Have a look at the whole file here: https://github.com/oolongbrothers/ansible/blob/97ac821b77510b9857d3ccdbe3f1443b0d5c35f5/lib/ansible/modules/packaging/os/flatpak_remote.py

**Comment by dagwieers** [comment]
I don't know why you don't work through this PR instead. It's impossible to comment on the file you referenced, and the individual commits do not show everything.

**Comment by oolongbrothers** [comment]
Sorry for the inconvenience, @dagwieers. Good point, that would be a much smoother way to work.

I don't have permissions to push to JayKayy's fork though. Maybe you could give me push access to your `flatpak` and `flatpak_repo` branches, @JayKayy ?

Or is there another smooth way to collaborate on a PR together?

**Comment by dagwieers** [comment]
@oolongbrothers If you create a PR based on a specific branch, you can update and collaborate on that branch and this PR will automatically follow that branch. So in order to collaborate on this PR, simply work together on branch https://github.com/JayKayy/flatpak-repo

**Comment by oolongbrothers** [comment]
All right! I tried pushing to the branch in JayKayy's fork that the PR is based on, but I don't have permissions.
In that case, I'll have to wait until @JayKayy gives me push access to the branch (and the `flatpak` one).

**Comment by oolongbrothers** [comment]
Ok, I did some changes to address your latest feedback, @dagwieers 

Please bear with us and our messy branches :innocent: . Just to get on with it until I would get push access to the PR branch, I did the following:
I created a new branch in my fork based on current upstream devel and this branch. Then I squashed all my changes into one commit: https://github.com/oolongbrothers/ansible/commit/c400fa190456764056e508fef90cef40881d233f. So you should be able to comment on anything, if you have more feedback :sweat_smile: .

**Comment by dagwieers** [comment]
Or we have to move this over to another PR and close this one, which may be easier ?

**Comment by oolongbrothers** [comment]
Right. I would normally prefer to have @JayKayy on board with everything, but since we are in a bit of a hurry and him seeming busy right now, this might be the only option to get things into shape in time.

@dagwieers Would you prefer that I retain all the historical commits, (because of our mess they are not complete though) or is it ok to quash everything into one commit (while of course duly crediting @JayKayy in the commit message)?

**Comment by dagwieers** [comment]
I don't need the individual commits, but what you could do is have the first commit include the changes from @JayKayy (in his name, Git allows this) and the second commit could be your changes. If you then make a PR from that branch against devel, I could make the (hopefully last) comments before merging.

**Comment by oolongbrothers** [comment]
Cool, working on it.

**Comment by JayKayy** [comment]
@oolongbrothers I believe i have given you access

**Comment by oolongbrothers** [comment]
Hi @JayKayy ! Sorry, must have been my bad then. I tried, but I may have done something wrong. I might have tried with the `flatpak` branch :(

Just before reading this I have created the following pull request: https://github.com/ansible/ansible/pull/40189
No intent to hijack your module, I just felt stuck getting things into shape before the freeze.

How should we proceed now? Should we proceed with this PR or switch to the new one.

**Comment by dagwieers** [comment]
@oolongbrothers Whatever is easiest. The other PR already has an issue with @mentions in the commit subject.

**Comment by oolongbrothers** [comment]
Yeah :/ saw it and attempted to fix it.

**Comment by oolongbrothers** [comment]
@JayKayy Maybe I'm doing something wrong. I still can't push to the `flatpak_repo` branch:
```
$ git checkout flatpak_repo 
Switched to branch 'flatpak_repo'
Your branch is up-to-date with 'origin/flatpak_repo'.
$ git push
ERROR: Permission to JayKayy/ansible.git denied to oolongbrothers.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Maybe we can stick to the new branch for this module now. But if you give me push access on the `flatpak` branch, we can continue working on getting that one up to speed tomorrow :smiley:

**Comment by JayKayy** [comment]
@oolongbrothers Sorry i have been away for a while. I would recommend taking the surest route to getting this functionality merged.

**Pull request closed without merging**
